AMQNET-575: NMS AMQP Client Rework

RagnarPaulson opened the pull request:
Add an NMS API implememntation that wraps the
AMQPnetLite .NET API.

Signed-off-by: Ragnar Paulson <Ragnar.Paulson@Solace.com>

gemmellr commented:
I cant really help on the code side as .NET isn't an area I play, but from a skim over the file list for curiosity it seemed like there are various non-code elements in this PR which perhaps shouldn't be included?

RagnarPaulson commented:
Good catch.  In my eagerness to compress several commits into one pull request I lost the .gitignore and inadvertently committed all the binaries/output.  I've removed them all now.

tabish121 commented:
It is generally best practice to squash commits into one on a PR to make life easier for the reviews and to make the commit history simpler.  We don't really need to preserve a commit removing files that shouldn't have been included in the first commit for instance.

RagnarPaulson commented:
I've tried to follow the quash and rebase instructions outlined https://github.com/servo/servo/wiki/Beginner's-guide-to-rebasing-and-squashing

But it's not clear to me that this made any difference to the pull request, quite possibly made it worse.

tabish121 commented:
If you can't work out the squash then another way to fix things is to back out the commit using a "git reset --soft HEAD~X" where 'X' is the number of commits and then create a new commit from there and push -f to overwrite the PR branch.

RagnarPaulson commented:
Thanks.  That seemed to work better.  Shows up as one commit now in the pull request.

tabish121 commented on HelloWorld/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
```
Missing License header here.

tabish121 commented on NMS.AMQP/NMS-AMQP.csproj:
```diff
@@ -0,0 +1,124 @@
+    <ProjectGuid>{0D8CF699-9702-4EC6-8719-C2968D32F09A}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>NMS.AMQP</RootNamespace>
```
The project should carry the Apache name e.g.  Apache.NMS.AMQP

tabish121 commented on NMS.AMQP/NMS-AMQP.csproj:
```diff
@@ -0,0 +1,124 @@
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>NMS.AMQP</RootNamespace>
+    <AssemblyName>NMS.AMQP</AssemblyName>
```
The assembly should also carry the Apache name.

tabish121 commented on NMS.AMQP/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,37 @@
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("NMS-AMQP")]
```
Assembly should carry the Apache name Apache-NMS-AMQP

tabish121 replied:
The assembly should carry the Apache name "Apache-NMS-AMQP"

tabish121 commented on NMS.AMQP/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,37 @@
+﻿using System.Reflection;
```
License header missing

tabish121 replied:
Missing license header.

tabish121 commented on README.md:
```diff
@@ -0,0 +1,112 @@
+# NMS-AMQP
```
Apache-NMS-AMQP

tabish121 replied:
Should be Apache NMS AMQP

tabish121 commented on src/main/csharp/Connection.cs:
```diff
@@ -0,0 +1,1158 @@
+using System.Reflection;
+using Apache.NMS.Util;
+
+namespace NMS.AMQP
```
Use namespace Apache.NMS.AMQP for all source in this library

tabish121 commented on NMS.AMQP/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,37 @@
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Apache")]
+[assembly: AssemblyProduct("NMS-AMQP")]
```
Same as above, add Apache to these names

tabish121 commented on NMS.AMQP/App.config:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
missing license header

tabish121 commented on src/main/csharp/Connection.cs:
```diff
@@ -0,0 +1,1158 @@
+            frame.IdleTimeOut = Convert.ToUInt32(connInfo.idleTimout);
+            frame.DesiredCapabilities = new Amqp.Types.Symbol[] {
+                SymbolUtil.OPEN_CAPABILITY_SOLE_CONNECTION_FOR_CONTAINER,
+                SymbolUtil.OPEN_CAPABILITY_DELAYED_DELIVERY
```
If allowing anonymous producers the client should be adding the ANONYMOUS-RELAY capability here as well

RagnarPaulson commented:
Thanks Tim,  All good feedback.   

What is the process now?  Should I commit my changes as a new commit, or quash it back into the original commit?

Ragnar

tabish121 commented:
I would just make the changes and amend the commit and force push it to the PR branch, you can also commit things locally and squash later if you prefer that route.  Basically until this goes into the project tree there isn't a need to preserve these intermediate code changes in the history

RagnarPaulson commented:
Got it.  I've just pushed up the latest version.  It still passes all the tests using ActiveMQ as a broker, so I believe i got all the namespace changes correct.

dpauls commented:
@tabish121 I appreciate the effort you've put into reviewing this.  Just a quick ping to see if there's anything else we can do to get this PR merged in.  If you just need to take another look and haven't had time yet, then no worries.  We're just eager to see this move towards a release.  Thanks!

tabish121 commented on src/main/csharp/ConnectionMetaData.cs:
```diff
@@ -0,0 +1,213 @@
+namespace Apache.NMS.AMQP
+{
+    /// <summary>
+    /// NMS.AMQP.ConnectionMetaData implements Apache.NMS.IConnectionMetaData
```
I spotted a few places where the API docs have the wrong package names based on the latest rounds of updates.  Since it isn't a real reference to a type it isn't as bad as it wouldn't break any doc generator but would be nice for the reader if they all used the right name.

tabish121 commented on src/example/csharp/HelloWorld/HelloWorld.cs:
```diff
@@ -0,0 +1,331 @@
+            Console.WriteLine("scheme: {0}", providerUri.Scheme);
+
+            StringDictionary properties = new StringDictionary();
+            //properties["clientId"] = "guest";
```
Some of these commented out bits in the example make it a bit less useful as a clean sample, this client Id and the one below here seem confusing

tabish121 commented on src/example/csharp/HelloWorld/HelloWorld.cs:
```diff
@@ -0,0 +1,331 @@
+                properties["nms.password"] = opts.password;
+            if (opts.clientId != null)
+                properties["NMS.CLIENTID"] = opts.clientId;
+            //properties["nms.clientid"] = "myclientid1";
```
See above

tabish121 commented on src/example/csharp/HelloWorld/HelloWorld.cs:
```diff
@@ -0,0 +1,331 @@
+            try
+            {
+
+
```
There's a decent amount of scattered extra white space in here that could be cleaned up while making other changes

tabish121 commented on src/example/csharp/HelloWorld/HelloWorld.cs:
```diff
@@ -0,0 +1,331 @@
+                prod.TimeToLive = TimeSpan.FromSeconds(20);
+                ITextMessage msg = prod.CreateTextMessage("Hello World!");
+
+                //IMapMessage msg = prod.CreateMapMessage();
```
Perhaps save these for other examples that show how to use the message types, adds extra confusion here for the new user

tabish121 commented on src/example/csharp/HelloWorld/HelloWorld.cs:
```diff
@@ -0,0 +1,331 @@
+                    //msg.Body.SetString("mykey", "Hello World! n:" + i);
+                    //msg.Body.SetBytes("myBytesKey", new byte[] { 0x65, 0x66, 0x54, (byte)(i & 0xFF) });
+
+                    // Stream  Msg Body
```
Same as above, this feels like it detracts from the simple HellowWorld example experience.

RagnarPaulson commented:
Thanks Tim, All good feedback.  I've cleaned up the comments and white space.  Added a new sample, and improved some of the 'toString' methods used in debugging/development.
